crates/bridge-exec/src/ws_client.rs

```diff
@@ -0,0 +1,159 @@
+//! Wrapper for managing a WebSocket client that allows recycling or restarting
+//! the client without restarting the entire program.
+
+use core::fmt;
+
+use deadpool::managed::{self, Manager, RecycleError, RecycleResult};
+use jsonrpsee::{
+    core::{
+        async_trait,
+        client::{BatchResponse, ClientT},
+        params::BatchRequestBuilder,
+        traits::ToRpcParams,
+        BoxError, ClientError, DeserializeOwned,
+    },
+    ws_client::{WsClient as WebsocketClient, WsClientBuilder},
+};
+
+/// Configuration for the WebSocket client.
+///
+/// settings that are necessary to initialize and configure
+/// the WebSocket client, ie. URL of the WebSocket server.
+#[derive(Clone, Debug)]
+pub struct WsClientConfig {
+    /// The URL of the WebSocket server.
+    pub url: String,
+}
+
+/// Manager for creating and recycling WebSocket clients.
+///
+/// This struct manages the lifecycle of WebSocket clients, including creating
+/// new clients and determining whether existing clients are still valid.
+#[derive(Clone, Debug)]
+pub struct WsClientManager {
+    /// The configuration used to initialize WebSocket clients.
+    pub config: WsClientConfig,
+}
+
+/// Represents the state of a WebSocket client.
+///
+/// - `Working`: The client is connected and operational.
+/// - `NotWorking`: The client is not connected or has encountered an error.
+#[derive(Debug)]
+pub enum WsClientState {
+    /// The WebSocket client is connected and operational.
+    Working(WebsocketClient),
+
+    /// The WebSocket client is not connected or is in a failed state.
+    NotWorking,
+}
+
+/// Wrapper for the WebSocket client state.
+///
+/// This struct encapsulates the `WsClientState`, enabling unified management of
+/// both connected and failed client states.
+#[derive(Debug)]
+pub struct WsClient(WsClientState);
+
+/// Implements the `Manager` trait for managing WebSocket clients.
+///
+/// The `WsClientManager` provides methods to create new clients and recycle
+/// existing ones, ensuring that clients remain in a valid state.
+impl Manager for WsClientManager {
+    type Type = WsClient;
+    type Error = jsonrpsee::core::StringError;
+
+    /// Creates a new WebSocket client.
+    ///
+    /// Attempts to initialize a new WebSocket client using the provided configuration.
+    /// Returns a `WsClient` wrapped in a `Working` or `NotWorking` state depending on
+    /// the success of the operation.
+    async fn create(&self) -> Result<Self::Type, Self::Error> {
+        let client = WsClientBuilder::default()
+            .build(self.config.url.clone())
+            .await;
+        let bl = match client {
+            Ok(cl) => WsClientState::Working(cl),
+            Err(_) => WsClientState::NotWorking,
+        };
```

Comment:
If we try to create a new instance and it starts off as being `NotWorking` that should be an error right?  When do we update the state from `Working` to `NotWorking`?